package/MDAnalysis/lib/pkdtree.py

```diff
@@ -0,0 +1,182 @@
+# D. L. Dotson, J. Domanski, S. Buchoux, I. M. Kenney, and O. Beckstein.
+# MDAnalysis: A Python package for the rapid analysis of molecular dynamics
+# simulations. In S. Benthall and S. Rostrup editors, Proceedings of the 15th
+# Python in Science Conference, pages 102-109, Austin, TX, 2016. SciPy.
+#
+# N. Michaud-Agrawal, E. J. Denning, T. B. Woolf, and O. Beckstein.
+# MDAnalysis: A Toolkit for the Analysis of Molecular Dynamics Simulations.
+# J. Comput. Chem. 32 (2011), 2319--2327, doi:10.1002/jcc.21787
+#
+
+"""
+PeriodicKDTree --- :mod:`MDAnalysis.lib.pkdtree`
+===============================================================================
+
+This module contains a class to allow searches on a KDTree involving periodic
+boundary conditions.
+"""
+
+from __future__ import absolute_import
+from six.moves import range
+
+import numpy as np
+from Bio.KDTree import _CKDTree
+
+__all__ = ['PeriodicKDTree', ]
+
+
+class PeriodicKDTree(object):
+    """
+    Wrapper around Bio.KDTree._CKDTree to enable search with periodic boundary
+     conditions.
+
+    A tree is first constructed with the coordinates wrapped onto the central
+    cell. A query for neighbors around a center point is performed first by
+    wrapping the center point coordinates to the central cell, then generating
+    images of this wrapped center point and finally searching for neighbors
+    close to the images.
+
+    Only the necessary number of center point images is generated for each
+    case. For instance, if the wrapped center point lies well within the cell
+    and far away from the cell boundaries, there may be no need to generate
+    any center point image.
+    """
+
+    def __init__(self, box, bucket_size=10):
+        """
+        Parameters
+        ----------
+        box : array-like or ``None``, optional, default ``None``
+          Simulation cell dimensions in the form of
+          :attr:`MDAnalysis.trajectory.base.Timestep.dimensions` when
+          periodic boundary conditions should be taken into account for
+          the calculation of contacts.
+        bucket_size : int
+          Number of entries in leafs of the KDTree. If you suffer poor
+          performance you can play around with this number. Increasing the
+          `bucket_size` will speed up the construction of the KDTree but
+          slow down the search.
+        """
+        self.dim = 3  # Strict implementation for 3D
+        self.kdt = _CKDTree.KDTree(self.dim, bucket_size)
+        self.built = 0
+        if len(box) != self.dim:
+            raise Exception('Expected array of length {}'.format(self.dim))
+        self.box = np.copy(np.asarray(box))
+        # zero is the flag for no periodic boundary conditions
+        self.box = np.where(self.box == np.inf, 0.0, self.box)
+        self.box = np.where(self.box > 0.0, self.box, 0.0)
+        if not self.box.any():
+            raise Exception('No periodic axis found')
```

Comment:
The whole box checking code can be the following.

```python
box = np.asarray(box)
if box.shape != (6,):
    raise TypeError("Except box of shape (6,)  with [x, y, z, alpha, beta, gamma]")
# sometimes no box is given as `[0, 0, 0, 90, 90, 90]` in MDAnalysis. This is a invalid box we also have
# to guard against.
if np.any(box[:3] == np.inf) or np.any(box[:3] < 0) or np.allclose(box[:3], 0):
    raise ValueError("box dimensions must be positive and finite")
```

I can also later refactor that into a new function called `valid_box` or `check_box` that we can use in other places of the library. But with these checks you can safely assume that all box dimensions are bigger then 0.

Reply:
I just found [_box_check](https://github.com/MDAnalysis/mdanalysis/blob/develop/package/MDAnalysis/lib/distances.py#L125), so I guess I'll use this. This function does not check for positive values, though.